API update (#138)
decentralized-identity/bbs-signature: draft-bbs-signatures.md

### Messages and generators

> Throughout the operations of this signature scheme, each message that is signed is paired with a specific generator (point in G1). Specifically, if a generator `H_1` is multiplied with `msg_1` during signing, then `H_1` MUST be multiplied with `msg_1` in all other operations as well (signature verification, proof generation and proof verification).

Reviewer: ```suggestion
Throughout the operations of this signature scheme, each message that is signed is paired with a specific generator (point in G1). Specifically, if a generator `H_1` is multiplied with `msg_1` during signing, then `H_1` MUST be multiplied with `msg_1` in all other operations (signature verification, proof generation and proof verification).
```

Author: Done! Thanks!

```
signature = Sign(SK, PK, header, to_sign)

Inputs:

- SK (REQUIRED), an octet string of the form outputted by the KeyGen
                 operation.
- PK (REQUIRED), an octet string of the form outputted by the SkToPk
                 operation provided the above SK as input.
- header (OPTIONAL), an octet string containing context and application
                     specific information. If not supplied, it defaults
                     to an empty string.
- to_sign (OPTIONAL), a list of (generator, message) pairs where the
                      generator is a point in G1 and the message is a
                      non-zero scalar mod r. If not supplied, will
                      default to the empty array ("[]").

Parameters:

> - Ciphersuite_ID (REQUIRED), ASCII string. The unique ciphersuite ID.

Reviewer 2: Can remove the notion of (REQUIRED) here as all parameters are required

Author: I'm happy to make the change if people thing that to not be that big of an issue.

This operation computes a deterministic signature from a secret key (SK) and optionally over a header and or a vector of messages.

```
signature = Sign(SK, PK, header, [(H_1, msg_1), ..., (H_L, msg_L)])

Inputs:

- SK (REQUIRED), an octet string of the form outputted by the KeyGen operation.
- PK (REQUIRED), an octet string of the form outputted by the SkToPk operation provided the above SK as input.
- header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.

> - [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generators and message pairs corresponding to messages to be signed. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, it will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.

Reviewer 2: ```suggestion
- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generators and message pairs corresponding to messages to be signed. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, it will default to the empty array ("[]").
```

Do we need the last sentence here?

Author: Indeed we don't. Thanks! Updated.

This operation computes a deterministic signature from a secret key (SK) and optionally over a header and or a vector of messages.

```
signature = Sign(SK, PK, header, [(H_1, msg_1),..., (H_L, msg_L)])

Inputs:

- SK (REQUIRED), an octet string of the form outputted by the KeyGen operation.
- PK (REQUIRED), an octet string of the form outputted by the SkToPk operation provided the above SK as input.
- header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.

> - [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list with pairs of the form (point in G1, scalar). Generators and messages pairs, corresponding to messages to be signed. If not supplied, it will default to the empty array ("[]"). The list MUST be ordered based on the values of the generators  (i.e., H_1 < H_2 < ... < H_L).

Reviewer 3: The list MUST be ordered based on the index of the generators?

Author: This does make sense considering how the generators are created but that order is harder to be "verified" unless the verifier re-creates all the generators (not really a security issue I think, it's just that the proof will not validate). 

Used the suggestion for now and opened issue #141 to discuss more.

This operation computes a deterministic signature from a secret key (SK) and optionally over a header and or a vector of messages.

```
signature = Sign(SK, PK, header, [(H_1, msg_1),..., (H_L, msg_L)])

Inputs:

- SK (REQUIRED), an octet string of the form outputted by the KeyGen operation.
- PK (REQUIRED), an octet string of the form outputted by the SkToPk operation 
     provided the above SK as input.
- header (OPTIONAL), an octet string containing context and application specific 
     information. If not supplied, it defaults to an empty string.
- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generators and message pairs 
     corresponding to messages to be signed. Generators must be G1 points, and the 

>      messages must be non-zero scalars mod q. If not supplied, it will default to the 

Reviewer 2: will need to be mod r if #140 is merged first

>      messages must be non-zero scalars mod q. If not supplied, it will default to the 

Reviewer 2: will need to be mod r if #140 is merged first

Definitions:

> - L, is the non-negative integer representing the number of messages to be signed i.e., length([(H_1, msg_1),..., (H_L, msg_L)]).  Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).

Reviewer 2: ```suggestion
- L, is the non-negative integer representing the number of messages to be signed i.e., length([(H_1, msg_1),..., (H_L, msg_L)]).  Note if the list of generators and message pairs are not supplied as an input to this operation, the value of L MUST evaluate to zero (0).
```

Author: Done.

This operation checks that a signature is valid for a given header and vector of messages against a supplied public key (PK).

```
result = Verify(PK, signature, header, [(H_1, msg_1),..., (H_L, msg_L)])

Inputs:

- PK (REQUIRED), an octet string of the form outputted by the SkToPk operation.
- signature (REQUIRED), an octet string of the form outputted by the Sign operation.
- header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.

> - [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generators and message pairs corresponding to signed messages. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, it will default to the empty array ("[]").

Reviewer 2: ```suggestion
- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generator and message pairs corresponding to signed messages. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, it will default to the empty array ("[]").
```

> - [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generator and message pairs. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, the list will default to the empty array ("[]"). The list MUST have the same order as when supplied to Sign.

Author: There is non! other than it is not "technically required", after the adoption of the signature domain.

That said, there may be use-cases where people want to use the un-linkable presentations part of bbs and not the selective showing one (like OAuth bearer tokens for example??)

                              1 to length(signed), in ascending order
                              (see above for requirements).

Parameters:

- Ciphersuite_ID (REQUIRED), ASCII string. The unique ciphersuite ID.
- H_s (REQUIRED), point of G1. The generator for the blinding value of
                  the signature.
- H_d (REQUIRED), point of G1. The generator used to sign the signature
                  domain.

Definitions:

- L, is the non-negative integer representing the number of signed
     messages i.e., L = length(signed). Note, if signed is not supplied
     as an input, the value of L MUST evaluate to zero (0).
- R, is the non-negative integer representing the number of revealed
     messages i.e., R = length(revealedIndexes). If revealedIndexes is
     not supplied as an input, the value of R MUST evaluate to zero (0).
- U, is the non-negative integer representing the number of un-revealed
     messages i.e., U = L - R.

Outputs:

- proof, octet string; or INVALID.

Procedure:

1. if length(revealedIndexes) > L, return INVALID

2. [i1, ..., iR] = revealedIndexes

3. [j1, ..., jU] = (1, ..., L) \ revealedIndexes

> 4. (H_i1, msg_i1), ..., (H_iR, msg_iR) = signed[i1], ..., signed[iR]

Reviewer 3: Currently unused, but I guess it will be when the revealed messages and indexes are hashed

7. if signature_result is INVALID, return INVALID

8. (A, e, s) = signature_result

9. if KeyValidate(PK) is INVALID, return INVALID

10. generators =  (H_s || H_d || H_1 || ... || H_L)

11. domain = hash_to_scalar((PK || L || generators || Ciphersuite_ID || header), 1)

12. (r1, r2, e~, r2~, r3~, s~) = hash_to_scalar(PRF(8*ceil(log2(r))), 6)

13. (m~_j1, ..., m~_jU) =  hash_to_scalar(PRF(8*ceil(log2(r))), U)

14. B = P1 + H_s * s + H_d * domain + H_1 * msg_1 + ... + H_L * msg_L

15. r3 = r1 ^ -1 mod r

16. A' = A * r1

17. Abar = A' * (-e) + B * r1

18. D = B * r1 + H_s * r2

19. s' = s + r2 * r3

20. C1 = A' * e~ + H_s * r2~

21. C2 = D * (-r3~) + H_s * s~ + H_j1 * m~_j1 + ... + H_jU * m~_jU

22. c = hash_to_scalar((PK || Abar || A' || D || C1 || C2 || ph), 1)

23. e^ = e~ + c * e

24. r2^ = r2~ + c * r2

25. r3^ = r3~ + c * r3

26. s^ = s~ + c * s'

27. for j in (j1, j2,..., jU): m^_j = m~_j + c * msg_j

28. proof = (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU))

29. return proof
```

### ProofVerify

This operation checks that a proof is valid for a header, vector of generators, revealed messages, and presentation header against a public key (PK).

> The generators are supplied as a list (called `generators`) of points in G1. The generators MUST be the same as the ones supplied to the [Sign](#sign) operation, as part of the (generator, message) pairs of the `to_sign` input list. The generators list MUST also be sorted based on the order that the (generator, message) pairs had in that list. More specifically, if the list `to_sign = [(H_1, msg_1), (H_2, msg_2), ..., (H_L, msg_L)]` was supplied to the Sign operation as input, then the generators list supplied as input to ProoVerify MUST equal `[H_1, H_2, ..., H_L]`, in that order.

Reviewer 3: *ProofVerify

> The revealed messages are supplied as a list of at most `length(generators)` pairs, of the form (revealed_message, index), containing a non-zero scalar mod r, representing a (revealed) message, and an integer in the 1 to L range, representing the index that the (generator, message) pair containing the revealed_message had in the `to_sign` list. The list MUST be sorted based on the (ascending) order of the indexes. If the list is not supplied, it will default to the empty array ("[]").

Reviewer 3: It seems a bit more natural to use (index, message) to me given that in many languages that makes the list naturally sortable, sorting by the first and then the second element

                     requirements).

Parameters:

- Ciphersuite_ID (REQUIRED), ASCII string. The unique ciphersuite ID.
- H_s (REQUIRED), point of G1. The generator for the blinding value of
                  the signature.
- H_d (REQUIRED), point of G1. The generator used to sign the signature
                  domain.

Definitions:

- L, is the non-negative integer representing the number of signed
     messages i.e., length(revealed). Note, if revealed is not supplied
     as an input, the value of L MUST evaluate to zero (0).
- R, is the non-negative integer representing the number of revealed
     messages i.e., R = length(revealedMsgs). If revealedMsgs is not
     suplied as an input, the value of R MUST evaluate to zero (0).
- U, is the non-negative integer representing the number of un-revealed
     messages i.e., U = L - R.

Outputs:

- result, either VALID or INVALID.

Procedure:

1. if length(revealedMsgs) > L, return INVALID

2. (msg_i1, i1), ..., (msg_iR, iR) = revealedMsgs

3. [j1, ..., jU] = (1, ..., L) \ [i1, ..., iR]

4. H_i1, ..., H_iR = generators[i1], ..., generators[iR]

5. H_j1, ..., H_jU = generators[j1], ..., generators[jU]

> 6. if KeyValidate(PK) is INVALID, return INVALID

Reviewer 3: This should probably remain near the beginning?

7. (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1,...,m^_jU)) = proof

> 8. generators =  (H_s || H_d || H_1 || ... || H_L)

Reviewer 3: Shadows the previous `generators` variable

Definitions:

> - L, is the non-negative integer representing the number of messages to be signed i.e., length([(H_1, msg_1),..., (H_L, msg_L)]). Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).

Reviewer 2: ```suggestion
- L, is the non-negative integer representing the number of messages to be signed i.e., length([(H_1, msg_1),..., (H_L, msg_L)]). Note if the list of generator and message pairs are not supplied as an input to this operation, the value of L MUST evaluate to zero (0).
```

### ProofGen

This operation computes a zero-knowledge proof-of-knowledge of a signature, while optionally selectively disclosing from the original set of signed messages. The "prover" may also supply a presentation header, see [presentation header selection](#presentation-header-selection) for more details. 

The generators and messages are supplied as pairs, along with a Boolean variable indicating if that message will be revealed or not. The generators and messages MUST be the ones supplied to the Sign operation.

```
proof = ProofGen(PK, signature, header, ph,
                 [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)])

Inputs:

- PK (REQUIRED), an octet string of the form outputted by the SkToPk operation.
- signature (REQUIRED), an octet string of the form outputted by the Sign operation.
- header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.
- ph (OPTIONAL), octet string.

> - [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting by a generator, the message and a variable indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.

Reviewer 2: ```suggestion
- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting of a generator, the message and a boolean value indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```
```suggestion
- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting by a generator, the message and a variable indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```

Reviewer 2: The last statement here around the ordering I still dont think is 100% clear, but I think in general this PR is a great improvement.

Reviewer 2: The last statement here around the ordering I still dont think is 100% clear, but I think in general this PR is a great improvement.

Author: Changed the statement a bit (now the ordering needs to follow the ordering of the input to Sign) and provided some examples to make it more clear.

> - [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting by a generator, the message and a variable indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.

Reviewer 2: ```suggestion
- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting of a generator, the message and a boolean value indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```
```suggestion
- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting by a generator, the message and a variable indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```

Author: Done! Thanks!

```
proof = ProofGen(PK, signature,
                 [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
                 [(H_h1, msg_h1), ..., (H_hU, msg_hU)],
                 header, ph)

Inputs:

- PK (REQUIRED), an octet string of the form outputted by the SkToPk operation.
- signature (REQUIRED), an octet string of the form outputted by the Sign operation.
- [(H_i1, msg_i1), ..., (H_iR, msg_iR)] (REQUIRED), Generators and message pairs 
                                        corresponding to messages to be revealed. 
                                        Generators must be G1 points, and the 
                                        messages must be non-zero scalars mod q.

> - [(H_h1, msg_h1), ..., (H_hU, msg_hU)] (REQUIRED), Generators and messages pairs 

Reviewer 3: Not sure why these two are required and can't default to [] as well

Author: My thinking was that if both lists are optional, it would allow both to be empty at the same time (which it should not if the original signed list is also non-empty). 

That said, I added a not in the description of ProofGen to highlight this fact and made the lists optional. I agree that this way is better from having the lists be required.

### ProofGen

This operation computes a zero-knowledge proof-of-knowledge of a signature, while optionally selectively disclosing from the original set of signed messages. The "prover" may also supply a presentation header, see [presentation header selection](#presentation-header-selection) for more details. 

The generators and messages are supplied as pairs in two different lists: the first list containing (generator, message) pairs containing messages that will be revealed, and the second list containing (generator, message) pairs corresponding to the signed messages that will be kept hidden. The union of those two lists MUST equal the list of (generator, message) pairs supplied to Sign (i.e., the second list MUST contain all (generator, message) pairs corresponding to signed messages that do not appear in the first list). This means that, although the two lists are not required (i.e., one can elect to not reveal any message etc.), both lists cannot be empty at the same time, assuming that the original list supplied to the Sign operation was also non-empty (if that's the case, then both lists must be empty)

```
proof = ProofGen(PK, signature, header, ph,

>                  [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 

Reviewer 2: [(H_1, msg_1, reveal?) ... (H_L, msg_L, reveal?)] an alternative syntax suggestion

>                  [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 

Reviewer 2: [(H_1, msg_1, reveal?) ... (H_L, msg_L, reveal?)] an alternative syntax suggestion

Definitions:

- L, is the non-negative integer representing the number of signed messages i.e., length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if the list of (generator, message boolean) tuples is not supplied, the value of L MUST evaluate to zero (0).
- R, is the non-negative integer representing the number of revealed messages i.e., R is the number of the boolean values: revealed_1, ..., revealed_L, that are "true". If no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).

> - U, is the non-negative integer representing the number of un-revealed messages i.e., U = L-R. If no non-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).

Reviewer: ```suggestion
- U, is the non-negative integer representing the number of un-revealed messages i.e., U = L-R. If no un-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).
```

Author: Thanks! Updated!

Definitions:

> - L, is the non-negative integer representing the number of signed messages e.g., length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if no messages are signed, the value of L MUST evaluate to zero (0).

Reviewer 2: ```suggestion
- L, is the non-negative integer representing the number of signed messages, length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if no messages are signed, the value of L MUST evaluate to zero (0).
```

Reviewer 2: Its not an example IMO so we should drop the e.g

> - R, is the non-negative integer representing the number of revealed messages e.g., R = revealed_1 + revealed_2 + ... + revealed_L. If no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).

Reviewer 2: ```suggestion
- R, is the non-negative integer representing the number of revealed messages e.g., R = revealed_1 + revealed_2 + ... + revealed_L. In the calculation of R its assumed that the boolean value of true evaluates to 1 whereas the boolean value of false evaluates to 0.  If no messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
```

Reviewer 2: Trying to clarify that effectively we are casting boolean values in the calculation of R to an integer?

Author: Good point! Updated to a different wording that it is sorter but may not be as clear?? If yes i will update to the suggestion.

> - U, is the non-negative integer representing the number of un-revealed messages e.g., U = L-R. If no non-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).

Reviewer 2: ```suggestion
- U, is the non-negative integer representing the number of un-revealed messages, U = L-R. If no non-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).
```

Definitions:

- L, is the non-negative integer representing the number of messages to be signed e.g length(msg_1,...,msg_L). Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).

> - R, is the non-negative integer representing the number of revealed messages e.g length(RevealedIndexes). Note if no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).H

Reviewer 3: RevealedIndexes doesn't exist anymore

Author: Thanks! Addressed.

1. signature_result = octets_to_signature(signature)

> 2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_i, msg_i), if revealed_i = 1]

Reviewer 2: I think this syntax needs to be explained more, in general we could elect to do more of a filter or map style function?

Reviewer 2: ```suggestion
2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_i, msg_i), if revealed_i = 1]
```
```suggestion
2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)].filter(item => item.revealed_i == true)
```

Reviewer 2: ```suggestion
2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_i, msg_i), if revealed_i = 1]
```
```suggestion
2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)].filter(item => item.revealed_i == true)
```

> 2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_i, msg_i), if revealed_i = 1]

Reviewer 2: I think this syntax needs to be explained more, in general we could elect to do more of a filter or map style function?

Author: I think that the complexity of having to define a filter function will outweigh the advantages of the current API??? In my last push I've described in "clear text" the operation of getting `(H_i1, msg_i1), ..., (H_iR, msg_iR)` etc., Thoughts on the alternative???

3. (A, e, s) = signature_result

4. if KeyValidate(PK) is INVALID, return INVALID

5. generators = (H_s || H_d || H_1 || ... || H_L)

6. domain = OS2IP(hash(PK || L || generators || Ciphersuite_ID || header)) mod q

7. for element in (r1, r2, e~, r2~, r3~, s~, m~_1, ..., m~_U):

8.      element = hash(PRF(8*ceil(log2(q)))) mod q

9.      if element = 0, go back to step 7

10. B = P1 + H_s * s + H_d * domain + H_1 * msg_1 + ... + H_L * msg_L

11. r3 = r1 ^ -1 mod q

12. A' = A * r1

13. Abar = A' * (-e) + B * r1

14. D = B * r1 + H_s * r2

15. s' = s + r2 * r3

16. C1 = A' * e~ + H_s * r2~

17. C2 = D * (-r3~) + H_s * s~ + H_h1 * m~_1 + ... + H_hU * m~_U

18. c = hash(PK || Abar || A' || D || C1 || C2 || ph)

19. e^ = e~ + c * e

20. r2^ = r2~ + c * r2

21. r3^ = r3~ + c * r3

22. s^ = s~ + c * s'

23. for j in (1, 2,..., U): m^_j = m~_j + c * msg_j

24. proof = (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU))

25. return proof
```

### ProofVerify

This operation checks that a proof is valid for a header, vector of revealed messages (along side their index corresponding to their original position when signed) and presentation header against a public key (PK).

```

> result = ProofVerify(PK, proof, header, ph,          
>                  [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
>                  [H_h1, ..., H_hU])

Reviewer 2: ```suggestion
result = ProofVerify(PK, proof, header, ph,          
                 [H_1, ..., H_L], 
                 [(index, msg_x), ..., (index, msg_x)])
```

Alternative syntax suggestion
```suggestion
result = ProofVerify(PK, proof, header, ph,          
                 [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
                 [H_h1, ..., H_hU])
```

> result = ProofVerify(PK, proof, header, ph,          
>                  [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
>                  [H_h1, ..., H_hU])

Reviewer 2: ```suggestion
result = ProofVerify(PK, proof, header, ph,          
                 [H_1, ..., H_L], 
                 [(index, msg_x), ..., (index, msg_x)])
```

Alternative syntax suggestion
```suggestion
result = ProofVerify(PK, proof, header, ph,          
                 [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
                 [H_h1, ..., H_hU])
```

Inputs:

- PK (REQUIRED), an octet string of the form outputted by the SkToPk operation.
- proof (REQUIRED), an octet string of the form outputted by the ProofGen operation.
- header (OPTIONAL), an optional octet string containing context and application 
     specific information. If not supplied, it defaults to an empty string.
- ph (OPTIONAL), octet string.
- [(H_i1, msg_i1), ..., (H_iR, msg_iR)] (OPTIONAL), a list of generators and message 
     pairs corresponding to revealed messages. Generators must be G1 points, and the 
     messages must be non-zero scalars mod q. If not supplied, it will default to the 
     empty array ("[]"). The list MUST be ordered based on the index of the 
     generators.

> - [H_h1, ..., H_hU] (OPTIONAL), a list of points in G1. Generators corresponding to 
>      messages to be kept hidden. If not supplied, the list will default to the empty 
>      array ("[]"). The list MUST be ordered based on the index of the generators.

Reviewer 2: ```suggestion
- [H_h1, ..., H_hU] (OPTIONAL), a list of points in G1. Generators corresponding to 
     un-revealed messages. If not supplied, the list will default to the empty 
     array ("[]"). The list MUST be ordered based on the index of the generators.
```

I think reveal vs un-reveal is better language than adding in another term "hidden"?

Author: Thanks! Addressed

Definitions:

- R, is the non-negative integer representing the number of revealed messages e.g.,
     length([(H_i1, msg_i1), ..., (H_iR, msg_iR)]). If no revealed messages are
     supplied as an input to this operation, the value of R MUST evaluate to zero (0).

> - U, is the non-negative integer representing the number of non-revealed messages e.g.,

Reviewer 2: ```suggestion
- U, is the non-negative integer representing the number of un-revealed messages e.g.,
```
